### PR TITLE
PO: Refactor Parsing

### LINKED ISSUE #282: Incorrect value extracted from PO
New line escaping in values is not correctly converted from the PO syntax into the Message Format 2 syntax.

```go
actual := "\n\\n\n                This filter was inherited from the dashboard's context.\\n\n                It won't be saved when saving the chart.\\n              "

expected := "\n                This filter was inherited from the dashboard's context.\n                It won't be saved when saving the chart.\n              "
```

**Message in PO**

```po
#: superset-frontend/src/explore/components/controls/DndColumnSelectControl/Option.tsx:71
#: superset-frontend/src/explore/components/controls/OptionControls/index.tsx:326
msgid ""
"\n"
"                This filter was inherited from the dashboard's context.\n"
"                It won't be saved when saving the chart.\n"
"              "
msgstr ""
```

**The message was extracted from the following source code (react)**
https://github.com/apache/superset/blob/4a1f66a6c74501dfc1f23b46a0c3e071251909bf/superset-frontend/src/explore/components/controls/OptionControls/index.tsx#L326

```javascript
            t(`
                This filter was inherited from the dashboard's context.
                It won't be saved when saving the chart.
              `)
```

Currently incorrectly converted to Message Format 2

```messageformat

\\n
                This filter was inherited from the dashboard's context.\\n
                It won't be saved when saving the chart.\\n
              
```

Expected text in Message Format 2

```messageformat

                This filter was inherited from the dashboard's context.
                It won't be saved when saving the chart.
              
```

**PO description**

> One should carefully distinguish between end of lines marked as ‘\n’ inside quotes, which are part of the represented string, and end of lines in the PO file itself, outside string quotes, which have no incidence on the represented string.
>
> ...external quotes are removed, multi-line strings are concatenated, and backslash escaped sequences are turned into their corresponding characters...
>
> https://www.gnu.org/software/gettext/manual/gettext.html

### LINKED ISSUE #284: No space or escaping new lines needed in `{{...}}`
- unnecessary space at the beginning of the text in `{{...}}`
- unexpected escaped new lines in`{{...}}`. New lines in `{{...}}` are the same new lines as in the original text, no need to escape.

In MF2, 

*Original text in source code*
```js
t('%(message)s\nThis may be triggered by: \n%(issues)s', {
        message,
        issues: extra.issue_codes
          .map(issueCode => issueCode.message)
          .join('\n'),
      })
```

*Actual*
```mf
.local $format = { python-format }
.local $message = { |%(message)s| }
.local $issues = { |%(issues)s| }
{{
{ $message }\\n
This may be triggered by: \\n
{ $issues }}}
```

*Expected*
```mf
.local $format = { python-format }
.local $message = { |%(message)s| }
.local $issues = { |%(issues)s| }
{{{ $message }
This may be triggered by: 
{ $issues }}}
```

### PR DESCRIPTION
**Pull request by VladislavsPerkanuks**
also closes #282  and closes #284

## Changes

- Nuked lexer, as it was complicating things more then it was helping
- Rewritten parser, more graceful + should be easier to modify if needed
- Refactor PO struct, removed the fields that are not used anywhere. If needed can be added
- Added integration test, where specific commit superset PO file is loaded, and tested against some known inputs (headers ,and 3 messages)
- moved PO marshal to separate file

## Benchmark
|                    | Old implementation      | New Implementation                                      |
| ------------------ | ------------ | -------------------------------------------- |
|                    | **sec/op**       | **sec/op     vs base**                           |
| _Parse/Superset-10 | 2293.9µ ± 1% | 541.7µ ± 1%        -76.39% (p=0.000 n=10+20) |
| _Parse/Odoo-10     | 458.8µ ± 3%  | 191.5µ ± 1%        -58.25% (p=0.000 n=10+20) |
| geomean            | 1.026m       | 322.1µ              -68.60%                  |

|                    | Old implementation      | New Implementation                                      |
| ------------------ | ------------- | -------------------------------------------- |
|                    | B/op          | B/op      vs base                            |
| _Parse/Superset-10 | 9.530Mi ± 0%  | 6.105Mi ± 0%       -35.94% (p=0.000 n=10+20) |
| _Parse/Odoo-10     | 1103.6Ki ± 0% | 783.8Ki ± 0%       -28.98% (p=0.000 n=10+20) |
| geomean            | 3.205Mi       | 2.162Mi            -32.55%                   |

|                    | Old implementation     | New Implementation                              |
| ------------------ | ----------- | ------------------------------------ |
|                    | allocs/op   | allocs/op   vs base                  |
| _Parse/Superset-10 | 46.73k ± 0% | 11.79k ± 0%        -74.77% (n=10+20) |
| _Parse/Odoo-10     | 8.407k ± 0% | 2.281k ± 0%        -72.87% (n=10+20) |
| geomean            | 19.82k      | 5.186k             -73.84%           |

### CODE REVIEW

**Review by kaaposc — APPROVED** [review]

**Comment by janishorsts** [comment]
In what sense lexer was not working when lexing character by character?

**Comment by VladislavsPerkanuks** [comment]
Removing Regex yields minimum improvements in benchmarks, and results in more nested/complex code

Metric | with_regex.txt | without_tegex.txt | Comparison vs base
 -- | -- | -- | --
 sec/op | 1.727m ± 9% | 1.653m ± 3% | -4.25% (p=0.001 n=10)
 B/op | 3.934Mi ± 0% | 3.851Mi ± 0% | -2.11% (p=0.000 n=10)
allocs/op | 10.65k ± 0% | 10.56k ± 0% | -0.89% (p=0.000 n=10)

File: pkg/convert/po.go [hunk 1]
```diff
@@ -19,28 +18,37 @@ import (
...
-	portableObject, err := po.Parse(bytes.NewReader(b))
+	portableObject, err := po.Parse(b)
 	if err != nil {
 		return model.Translation{}, fmt.Errorf("parse portableObject file: %w", err)
 	}
 
+	var lang language.Tag
+
+	if portableObject.Headers.Get("Language") != "" {
+		lang, err = language.Parse(portableObject.Headers.Get("Language"))
```
Comment (kaaposc) [inline pkg/convert/po.go:29]: Alternative
```suggestion
	if l := portableObject.Headers.Get("Language"); l != "" {
		lang, err = language.Parse(l)
```

**Review by DarjaLeonova — APPROVED** [review]

File: pkg/po/po.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package po
+
+type PO struct {
+	Headers  Headers
+	Messages []Message
+}
+
+type Headers []Header
```
Comment (VladislavsPerkanuks) [inline pkg/po/po.go:8]: slice instead map, to preserve order

**Review by janishorsts — APPROVED** [review]
I did not review it. Please, merge!

**Pull request merged**